python/pyarrow/array.pxi

```diff
@@ -1170,7 +1170,9 @@ cdef class Array(_PandasConvertible):
         array = PyObject_to_object(out)
 
         if isinstance(array, dict):
+            missings = array["indices"] < 0
             array = np.take(array['dictionary'], array['indices'])
+            array[missings] = np.NaN
```

Comment:
I thought about adding a `if missings.any():` here, but it didn't change performances at all, so not much benefit.

```
def withoutcheck():
    arr[mask] = np.NaN

def withcheck():
    if mask.any():
        arr[mask] = np.NaN

>>> timeit.timeit(withoutcheck, number=10000)
5.555068847000001
>>> timeit.timeit(withcheck, number=10000)
5.6257486310000075
```

Reply:
Checking the null count of the original arrow array (`self`) might help?

Reply:
I noticed I also have to check for interaction with zero copy. Given that we are ending up _writing_ to `array[missings]` in the end.

Reply:
Added an optimization based on `zero_copy_only` option (as it doesn't allow nulls) and on `self.null_count` as the null_count is cached ( https://github.com/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/cpp/src/arrow/array/data.cc#L120-L121 ) thus should frequently not add to the overhead.

Also confirmed that `-1` is used to signal NULL values when converting to `numpy` arrays ( https://github.com/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/cpp/src/arrow/python/arrow_to_pandas.cc#L1637 )

A further optimization might have been to have `ConvertArrayToPandas`  return the count of null values (as it is invoking `IsValid` on them anyway) but that requires a more widespread change and thus I think should be deferred until proved necessary.